### PR TITLE
Fix iterkeys(), iteritems(), itervalues()

### PR DESCRIPTION
**Pull request by cclauss**
Four methods of __dict__ that were removed in Python 3.  Use dict and set comprehensions where useful.

### CODE REVIEW

File: infogami/core/helpers.py [hunk 1]
```diff
@@ -22,7 +22,7 @@ def flatten(d):
     {'a': 1, 'b#0': 2, 'b#1': 3, 'c.x': 4, 'c.y': 5}
     """
     def traverse(d, prefix, delim, visit):
-        for k, v in d.iteritems():
+        for k, v in d.items():
```
Comment (hornc) [inline infogami/core/helpers.py]: @cclauss  I'm worried this change might cause a performance issue under Python 2, and maybe also in Python 3, I don't _think_ Python 3 `dict.items()` returns an iterator by default? Actually (after some forum searching), I'm not sure about 3, I think the returned `dict_items([])` in 3 will behave like an iterator if used as one... either way, it does change the behaviour under Python 2.

 A more backwards compatible change might be:
`for k, v in iter(d.items()):`   which I _think_ does the same thing in Python 3 and 2.

Thoughts?
If this is a better change, it'd apply to all changed lines in this PR.

Python 2.7 

```
>>> a = {}
>>> a.items()
[]
>>> a.iteritems()
<dictionary-itemiterator object at 0x7f185c096998>
>>> iter(a.iteritems())
<dictionary-itemiterator object at 0x7f185c0969f0>
>>> iter(a.items())
<listiterator object at 0x7f185c0aa050>
```

Python 3
```
>>> a = {}
>>> a.items()
dict_items([])
>>> iter(a.items())
<dict_itemiterator object at 0x7f7cdf156098>
```

Comment (cclauss) [reply infogami/core/helpers.py]: There are three basic approaches to this:
1. list(a.items())  Slows down Python 3 but is preferred by Python-modernize
2. iter(a.items()) Your suggestion but Python 2 is still going to build the list in the inner block so there is no performance or memory footprint gain and the result can only be used once
3. The way that __six__ does it https://github.com/benjaminp/six/blob/master/six.py#L581

3. is clearly the best and delivers equivalent functionality in both 2 and 3.

File: infogami/core/helpers.py [hunk 2]
```diff
@@ -57,11 +57,11 @@ def setdefault(d, k, v):
         elif '#' in k:
             a, b = k.split('#', 1)
             return setdefault(setdefault(d, a, betterlist()), b, v)
-        else: 
+        else:
             return d.setdefault(k, v)
 
     d2 = {}
-    for k, v in d.iteritems():
+    for k, v in d.items():
```
Comment (cclauss) [inline infogami/core/helpers.py]: Used only once.  Python 2 will build a list in memory which will take up more RAM.  Python 3 will be fast.

File: infogami/core/helpers.py [hunk 3]
```diff
@@ -108,7 +108,7 @@ def trimlist(x):
 
     def trimdict(x):
         y = {}
-        for k, v in x.iteritems():
+        for k, v in x.items():
```
Comment (cclauss) [inline infogami/core/helpers.py]: Used only once.  Python 2 will build a list in memory which will take up more RAM.  Python 3 will be fast.

File: infogami/infobase/_dbstore/save.py [hunk 4]
```diff
@@ -97,7 +97,7 @@ def index(key, value):
                 for v in value:
                     index(key, v)
 
-        for k, v in data.iteritems():
+        for k, v in data.items():
```
Comment (cclauss) [inline infogami/infobase/_dbstore/save.py]: Used only once.  Python 2 will build a list in memory which will take up more RAM.  Python 3 will be fast.

File: infogami/infobase/_dbstore/save.py [hunk 5]
```diff
@@ -366,7 +366,7 @@ def compile_index(self, index):
         """Compiles doc-index into db-index.
         """
         keys = set(key for type, key, datatype, name in index)
-        for (type, key, datatype, name), values in index.iteritems():
+        for (type, key, datatype, name), values in index.items():
```
Comment (cclauss) [inline infogami/infobase/_dbstore/save.py]: Used only once.  Python 2 will build a list in memory which will take up more RAM.  Python 3 will be fast.

File: infogami/infobase/_dbstore/save.py [hunk 6]
```diff
@@ -386,7 +386,7 @@ def get_pid(type, name):
 
         dbindex = {}
 
-        for (type, key, datatype, name), values in index.iteritems():
+        for (type, key, datatype, name), values in index.items():
```
Comment (cclauss) [inline infogami/infobase/_dbstore/save.py]: Used only once.  Python 2 will build a list in memory which will take up more RAM.  Python 3 will be fast.

File: infogami/infobase/_dbstore/save.py [hunk 7]
```diff
@@ -399,15 +399,15 @@ def group_index(self, index):
         """Groups the index based on table.
         """
         groups = defaultdict(dict)
-        for (table, thing_id, property_id), values in index.iteritems():
+        for (table, thing_id, property_id), values in index.items():
```
Comment (cclauss) [inline infogami/infobase/_dbstore/save.py]: Used only once.  Python 2 will build a list in memory which will take up more RAM.  Python 3 will be fast.

File: infogami/infobase/_dbstore/save.py [hunk 8]
```diff
@@ -399,15 +399,15 @@ def group_index(self, index):
...
             groups[table][thing_id, property_id] = values
         return groups
 
     def ignore_long_values(self, index):
         """The DB schema has a limit of 2048 chars on string values. This function ignores values which are longer than that.
         """
         is_too_long = self._is_too_long
-        return dict((k, [v for v in values if not is_too_long(v)]) for k, values in index.iteritems())
+        return {k: [v for v in values if not is_too_long(v)] for k, values in index.items()}
```
Comment (cclauss) [inline infogami/infobase/_dbstore/save.py]: Used only once.  Python 2 will build a list in memory which will take up more RAM.  Python 3 will be fast.

File: infogami/infobase/_dbstore/save.py [hunk 9]
```diff
@@ -332,7 +332,7 @@ def _dict_difference(self, d1, d2, eq=None):
         """Returns set equivalant of d1.difference(d2) for dictionaries.
         """
         eq = eq or (lambda a, b: a == b)
-        return dict((k, v) for k, v in d1.iteritems() if not eq(v, d2.get(k)))
+        return {k: v for k, v in d1.items() if not eq(v, d2.get(k))}
```
Comment (cclauss) [inline infogami/infobase/_dbstore/save.py]: Used only once.  Python 2 will build a list in memory which will take up more RAM.  Python 3 will be fast.

File: infogami/core/helpers.py (hunk 1 above)
Comment (cclauss) [reply infogami/core/helpers.py]: % __python2.7__
```python
import sys
>>> a = {key: value for value, key in enumerate("abcdefghijklmnopqrstuvwxyz")}
>>> a
{'a': 0, 'c': 2, 'b': 1, 'e': 4, 'd': 3, 'g': 6, 'f': 5, 'i': 8, 'h': 7, 'k': 10, 'j': 9, 'm': 12, 'l': 11, 'o': 14, 'n': 13, 'q': 16, 'p': 15, 's': 18, 'r': 17, 'u': 20, 't': 19, 'w': 22, 'v': 21, 'y': 24, 'x': 23, 'z': 25}
>>> sys.getsizeof(a.items())
280
>>> sys.getsizeof(iter(a.items()))
64
```

Comment (hornc) [reply infogami/core/helpers.py]: @cclauss Let's go for 3. and use `six` -- I just canvased @mekarpeles and that is his preference too.

Very good point about my 2. suggestion not making anything better, I clearly didn't think that through properly before typing :)

**Pull request merged**